### PR TITLE
Add Movsx() and Movzx() + Tests

### PR DESCRIPTION
**Pull request by tuday2**

### CODE REVIEW

File: MBBSEmu/CPU/CPUCore.cs [hunk 1]
```diff
@@ -944,7 +950,7 @@ private ushort GetOperandOffset(OpKind opKind)
         }
 
         /// <summary>
-        ///     Returns if the Current Instruction is an 8-bit or 16-bit operation
+        ///     Returns if the Current Instruction is an 8-bit, 16-bit, or 32-bit operation
```
Comment (tuday2) [inline MBBSEmu/CPU/CPUCore.cs:953]: Fixed Typo

File: MBBSEmu/CPU/CPUCore.cs [hunk 2]
```diff
@@ -2500,7 +2506,7 @@ private void Op_Push()
                     Push(GetOperandValueUInt8(_currentInstruction.Op0Kind, EnumOperandType.Destination));
                     return;
 
-                //166bit, also default size (-1)
+                //16bit, also default size (-1)
```
Comment (tuday2) [inline MBBSEmu/CPU/CPUCore.cs:2509]: Fixed Typo

File: MBBSEmu/CPU/CPUCore.cs [hunk 3]
```diff
@@ -3626,6 +3632,9 @@ private void Op_Lodsw()
             }
         }
 
+        /// <summary>
+        ///     Move data from String to String
+        /// </summary>
```
Comment (tuday2) [inline MBBSEmu/CPU/CPUCore.cs:3637]: Added Description

File: MBBSEmu/CPU/CPUCore.cs [hunk 4]
```diff
@@ -3651,6 +3660,120 @@ private void Op_Movsw()
...
+
+        /// <summary>
+        ///     Move with Sign-Extend to 32bit destination
+        /// </summary>
+        [MethodImpl(CompilerOptimizations)]
+        private uint Op_Movsx_32()
+        {
+            var sourceSize = 0;
+
+            if (_currentInstruction.Op1Kind == OpKind.Register)
```
Comment (enusbaum) [inline MBBSEmu/CPU/CPUCore.cs:3698]: Could these be done using a switch expression like the result below?

`var sourceSize = _currentInstruction.Op1Kind switch`

File: MBBSEmu/CPU/CPUCore.cs [hunk 5]
```diff
@@ -4075,5 +4198,43 @@ private void Flags_EvaluateSignZero(uint result)
...
+        /// <summary>
+        ///     Returns the size of the given Register
+        /// </summary>
+        /// <param name="register"></param>
+        /// <returns></returns>
+        public byte GetSize(Register register)
+        {
+            return register switch
+            {
+                Register.AL => 8,
```
Comment (enusbaum) [inline MBBSEmu/CPU/CPUCore.cs:4211]: As I look at these, let's change these constants to `sizeof(byte)`, `sizeof(ushort)`, `sizeof(uint)` just so people know what they are.

**Review by enusbaum — CHANGES_REQUESTED** [review]

File: MBBSEmu/CPU/CPUCore.cs (hunk 4 above)
Comment (tuday2) [reply MBBSEmu/CPU/CPUCore.cs]: yup! cleaned it up on both sx/zx and switched to 1 = 8bit, 2 = 16bit for consistency

File: MBBSEmu/Extensions/ByteExtensions.cs [hunk 6]
```diff
@@ -63,5 +63,35 @@ public static bool Parity(this byte b)
...
+        /// <summary>
+        ///     Sign extends 8bit -> 16bit
+        /// </summary>
+        /// <param name="b"></param>
+        /// <param name="bitDifference"></param>
+        /// <returns></returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ushort ToUshortSignExtended(this byte b)
+        {
+            if (b.IsNegative())
```
Comment (paladine) [inline MBBSEmu/Extensions/ByteExtensions.cs:76]: can be replaced with a simpler `return (ushort)(sbyte)b;`

File: MBBSEmu/Extensions/ByteExtensions.cs [hunk 7]
```diff
@@ -63,5 +63,35 @@ public static bool Parity(this byte b)
...
+        /// <param name="b"></param>
+        /// <param name="bitDifference"></param>
+        /// <returns></returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint ToUintSignExtended(this byte b)
+        {
+            if (b.IsNegative())
+                return (uint)(b | 0xFFFFFF00);
+
+            return (uint)b;
```
Comment (paladine) [inline MBBSEmu/Extensions/ByteExtensions.cs:94]: can be replaced with a simpler `return (uint)(sbyte)b;`

File: MBBSEmu/Extensions/UshortExtensions.cs [hunk 8]
```diff
@@ -67,5 +67,20 @@ public static bool Parity(this ushort b)
...
+        /// <param name="b"></param>
+        /// <param name="bitDifference"></param>
+        /// <returns></returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static uint ToUintSignExtended(this ushort b)
+        {
+            if (b.IsNegative())
+                return (uint)(b | 0xFFFF0000);
+
+            return (uint)b;
```
Comment (paladine) [inline MBBSEmu/Extensions/UshortExtensions.cs:83]: can be replaced with a simpler `return (uint)(short)b;`

**Review by paladine — APPROVED** [review]

File: MBBSEmu/Extensions/UshortExtensions.cs (hunk 8 above)
Comment (tuday2) [reply MBBSEmu/Extensions/UshortExtensions.cs]: Updated all 3 with wizardry!

**Review by enusbaum — APPROVED** [review]

**Pull request merged**